src/dht_proxy_server.cpp

```diff
@@ -0,0 +1,397 @@
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->getNodeInfo(session);
+            }
+        );
+        service_->publish(resource);
+        resource = std::make_shared<restbed::Resource>();
+        resource->set_path("/{hash: .*}");
+        resource->set_method_handler("GET",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->get(session);
+            }
+        );
+        resource->set_method_handler("LISTEN",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->listen(session);
+            }
+        );
+        resource->set_method_handler("POST",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->put(session);
+            }
+        );
+#if OPENDHT_PROXY_SERVER_IDENTITY
+        resource->set_method_handler("SIGN",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->putSigned(session);
+            }
+        );
+        resource->set_method_handler("ENCRYPT",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->putEncrypted(session);
+            }
+        );
+#endif // OPENDHT_PROXY_SERVER_IDENTITY
+        service_->publish(resource);
+        resource = std::make_shared<restbed::Resource>();
+        resource->set_path("/{hash: .*}/{value: .*}");
+        resource->set_method_handler("GET",
+            [this](const std::shared_ptr<restbed::Session> session)
+            {
+                this->getFiltered(session);
+            }
+        );
+        service_->publish(resource);
+
+        // Start server
+        auto settings = std::make_shared<restbed::Settings>();
+        settings->set_default_header("Content-Type", "application/json");
+        std::chrono::milliseconds timeout(std::numeric_limits<int>::max());
+        settings->set_connection_timeout(timeout); // there is a timeout, but really huge
+        settings->set_port(port);
+        try {
+            service_->start(settings);
+        } catch(std::system_error& e) {
+            // Fail silently for now.
+        }
+    });
+}
+
+DhtProxyServer::~DhtProxyServer()
+{
+    stop();
+}
+
+void
+DhtProxyServer::stop()
+{
+    service_->stop();
+    // listenThreads_ will stop because there is no more sessions
+    for (auto& listenThread: listenThreads_)
+        if (listenThread->joinable())
+            listenThread->join();
+    if (server_thread.joinable())
+        server_thread.join();
+}
+
+void
+DhtProxyServer::getNodeInfo(const std::shared_ptr<restbed::Session>& session) const
+{
+    const auto request = session->get_request();
+    int content_length = std::stoi(request->get_header("Content-Length", "0"));
+    session->fetch(content_length,
+        [&](const std::shared_ptr<restbed::Session> s, const restbed::Bytes& b)
```

Comment:
don't capture by reference in async operations

Reply:
done. my bad